### ISSUE TITLE
(Design suggestion) Use the winit crate instead of directly using X11/Wayland

### ISSUE THREAD

**Issue by hardBSDk** [wez/wezterm#4553]
The winit crate supports more platforms as it runs on backends beyond X11 and Wayland.

**Comment by jokeyrhyme** [comment]
https://github.com/wez/wezterm/issues/296#issuecomment-708499143

> Regarding your comments about winit: when I started building wezterm I found that it had a number of deficiencies and that there's a lot about its event loop that isn't compatible with how I want to structure my project.

I wonder if winit's API has changed sufficiently or if the thinking around this has otherwise changed?

It looks like winit 0.29.x internally uses smithay-client-toolkit 0.18.x , so this might be a pathway to getting onto a newer version of sctk (seeing as both paths are non-trivial)

**Comment by wez** [comment]
At this point, moving to winit is a significant rewrite of the windowing layer, so it is not to be attempted lightly.
All of these things would be need to meaningfully better in winit than they are in wezterm in order for it to be worth the effort of migrating:

* winit never used to support opening more than 1 window at a time in a project. I hope that has been resolved by now.
* winit's key event processing is different and may not be as flexible or comprehensive as wezterm's. For a terminal emulator where the users are very demanding about keys, this could be a huge blocker. There is a significant following of Japanese users that appreciate the IME implementation in wezterm, for example.
* wezterm's use of xcb led to a very smooth and snappy startup compared to other frameworks when I first built it. It would be a shame to regress that.

Do no underestimate the cost of dealing with the inevitable bugs that will be introduced by such an effort. I think those would be enough to absorb all my available time for quite a while after making such a change.

At the end of the day, I don't think there is a huge amount gained simply by switching to winit: most of the heavy GUI stuff is done in opengl or webgpu anyway.  The main advantage would be in insulating wezterm from sctk and wayland crates, but even that won't help, as we'd still want to talk certain wayland protocols directly anyway.
